torch/distributed/_spmd/graph_optimization.py

```diff
@@ -0,0 +1,408 @@
+
+
+def graph_optimization_pass(
+    run_after: Iterable[str] = tuple(),
+) -> Callable[..., Callable[..., None]]:
+    """
+    The contract of graph optimization pass. All the passes should be wrapped with
+    this decorator.
+    """
+
+    def inner(
+        func: Callable[..., Union[fx.GraphModule, IterGraphModule]]
+    ) -> Callable[..., Union[fx.GraphModule, IterGraphModule]]:
+        for name in run_after:
+            _run_before_sets[name].add(func.__name__)
+
+        @wraps(func)
+        def pass_wrapper(
+            gm: Union[fx.GraphModule, IterGraphModule], *args: Any, **kwargs: Any
+        ) -> None:
+            assert isinstance(
+                gm, (fx.GraphModule, IterGraphModule)
+            ), "The first argument of the pass must be either fx.GraphModule or IterGraphModule"
+            assert (
+                func.__name__ not in _optimized_func
+            ), f"Cannot apply {func.__name__} twice."
+            invalid_passes = _run_before_sets[func.__name__].intersection(
+                _optimized_func
+            )
+            assert (
+                not invalid_passes
+            ), f"{invalid_passes} must run after {func.__name__}"
+
+            func(gm, *args, **kwargs)
+            gm.graph.lint()
+            gm.graph.eliminate_dead_code()
+            gm.recompile()
+            _optimized_func.add(func.__name__)
+
+            prefix = f"after_{func.__name__}"
+            if _dump_graph_folder:
+                if isinstance(gm, IterGraphModule):
+                    dump_graphs_to_files(
+                        {
+                            f"{prefix}_setup_gm": gm.setup_gm,
+                            f"{prefix}_main_gm": gm.main_gm,
+                            f"{prefix}_cleanup_gm": gm.cleanup_gm,
+                        },
+                        _dump_graph_folder,
+                    )
+                else:
+                    dump_graphs_to_files({prefix: gm}, _dump_graph_folder)
+
+        return pass_wrapper
+
+    return inner
+
+
+@dataclass(unsafe_hash=True)
+class CommBlock:
+    shape: Optional[torch.Size]
+    node_list: List[fx.Node]
+    inputs: List[fx.Node]
+    wait_nodes: List[fx.Node]
+    comm_node: fx.Node
+    outputs: Set[fx.Node]
+
+
+def get_comm_block_nodes(comm_node: fx.Node) -> CommBlock:
+    """
+    Given a wait_comm node, find out all the nodes belong to this communcation.
+
+    Args:
+        comm_node(fx.Node): The target communication/collective node.
+    Returns:
+        The CommBlock that encapsulates the related nodes (e.g., wait_node) of
+        the given comm_node.
+    """
+    node_list = []
+    wait_nodes = []
+    inputs, _ = tree_flatten((comm_node.args, comm_node.kwargs))
+    distance = 0
+    wait_prefixes = ("wait_comm", "wait_tensor")
+    non_end_users_nodes = ("split", "reshape", "getitem", "detach", "alias")
+
+    nodes = collections.deque([comm_node, None])
+    while nodes and distance < 5:
```

Comment:
any reason we pick 5 as the distance here?

Reply:
Oh I see, if this is the first pass after getting DTensor expanded graph, the wait should be right after allreduce as the latter is triggered on demand.

Reply:
But does this still hold for sparse? cc @yifuwang

Reply:
5 is just a safety guard based again on the expansion as you mentioned. However, even if there are optimization pass that move wait around, the distance should not change? The BFS directly traverse the users. So theoretically, the distance between comm_ops and wait_tensor is 1. 5 is used for wait_comm

Reply:
Ah, yep, you are right, this is BFS distance